### PR TITLE
Track whether an optimization decision was cost-based

### PR DESCRIPTION
**Pull request by mlyublena**
Some of Presto's optimizers are heuristic, while others are cost-based. This change allows tracking which optimizers were driven by a cost-based decision (independent of whether the cost was estimated or supplied by HBO). This information is added to PlanOptimizerInformation and can be seen in the explain plan when verbose_optimizer_info_enabled=true, for example:

```
presto:tpch> explain select lineitem.linenumber,count(*) from orders join lineitem on (lineitem.orderkey=orders.orderkey) group by linenumber;
                                                                                                                                                                                                                                        Query Plan
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[PlanNodeId 12][linenumber, _col1] => [linenumber:integer, count:bigint]
         _col1 := count (1:36)
...
 Triggered optimizers: [AddLocalExchanges, ApplyConnectorOptimization, HashGenerationOptimizer, PickTableLayoutWithoutPredicate, PruneJoinChildrenColumns, PruneJoinColumns, PruneProjectColumns, PruneTableScanColumns, PruneUnreferencedOutputs, PushPartialAggregationThroughExchange, RemoveRedundantDistinctAggregation, RemoveRedundantIdentityProjections, ReorderJoins, SetFlatteningOptimizer, SimplifyPlanWithEmptyInput, StatsRecordingPlanOptimizer, UnaliasSymbolReferences]
 Applicable optimizers: [AddNotNullFiltersToJoinNode, KeyBasedSampler, MergePartialAggregationsWithFilter, PushPartialAggregationThroughJoin]
 Cost-based optimizers: [PushPartialAggregationThroughExchange(CBO), ReorderJoins(CBO)]
```

Note: currently we don't track whether the cost came from HBO or through cost estimation, only that the decision was cost-driven. We already have a change that tracks what the source of the cost estimation is, so we can potentially intersect the two to find this information. I could potentially reconsider and track and log this information directly here.

## Description
<!---Describe your changes in detail-->

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Print information about cost-based optimizers and the source of stats they use (CBO/HBO) in explain plans when session property verbose_optimizer_info_enabled=true

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java [hunk 1]
```diff
@@ -82,12 +83,19 @@ public PushPartialAggregationThroughExchange(FunctionAndTypeManager functionAndT
                             .matching(node -> !node.getOrderingScheme().isPresent())
                             .capturedAs(EXCHANGE_NODE)));
 
+    private boolean isCostBased;
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:86]: line break

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 2]
```diff
@@ -201,6 +211,8 @@ private void addJoinsWithDifferentDistributions(JoinNode joinNode, List<PlanNode
 
     private JoinNode getSyntacticOrderJoin(JoinNode joinNode, Context context, JoinDistributionType joinDistributionType)
     {
+        isCostBased = false;
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:214]: I wonder if caller can pass this information rather setting this here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java [hunk 3]
```diff
@@ -300,18 +300,21 @@ private static String indentMultilineString(String string, int level)
...
+        List<String> hboOptimizerNames = planOptimizerInfo.stream().filter(x -> x.getIsCostBased().isPresent() && x.getIsCostBased().get()).map(x -> x.getOptimizerName()).distinct().sorted().collect(toList());
 
         String triggered = "Triggered optimizers: [" +
-                String.join(", ", triggeredOptimizers) + "]\n";
+                String.join(", ", triggeredOptimizerNames) + "]\n";
         String applicable = "Applicable optimizers: [" +
-                String.join(", ", applicableOptimizers) + "]\n";
-        return triggered + applicable;
+                String.join(", ", applicableOptimizerNames) + "]\n";
+        String hbo = "HBO optimizers: [" +
```
[the remark this answers is not in the document]
Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java:314]: good point, we can call them Cost based and add information about whether the source of the cost was HBO or heuristic estimates

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 4]
```diff
@@ -63,12 +64,20 @@ public class DetermineJoinDistributionType
     private final CostComparator costComparator;
     private final TaskCountEstimator taskCountEstimator;
 
+    // records whether distribution decision was cost-based
+    private boolean isCostBased;
```
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:68]: \n

File: presto-main/src/main/java/com/facebook/presto/sql/Optimizer.java [hunk 5]
```diff
@@ -177,9 +177,9 @@ private void collectOptimizerInformation(PlanOptimizer optimizer, PlanNode oldNo
                 isTriggered ||
                 !optimizer.isEnabled(session) && isVerboseOptimizerInfoEnabled(session) &&
                         optimizer.isApplicable(oldNode, session, TypeProvider.viewOf(variableAllocator.getVariables()), variableAllocator, idAllocator, warningCollector);
-
-        if (isTriggered || isApplicable) {
-            session.getOptimizerInformationCollector().addInformation(new PlanOptimizerInformation(optimizerName, isTriggered, Optional.of(isApplicable), Optional.empty()));
+        boolean isCostBased = optimizer.isCostBased(session);
+        if (isTriggered || isApplicable || isCostBased) {
```
[the remark this answers is not in the document]
Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/Optimizer.java:181]: I think you're right, I'll change it to only log isCostBased if the optimization triggered

**Comment by ClarenceThreepwood** [comment]
Add these as well?

DetermineSemiJoinDistributionType
TransformDistinctInnerJoinToRightEarlyOutJoin

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java (hunk 1 above)
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:86]: And move to top

**Review by pranjalssh — APPROVED** [review]
Looks good except the existing comments

**Pull request merged**